src/t_hash.c

```diff
@@ -42,6 +42,11 @@ void hashTypeTryConversion(robj *o, robj **argv, int start, int end) {
     size_t sum = 0;
 
     if (o->encoding != OBJ_ENCODING_LISTPACK) return;
+    size_t args_length = end - start + 1;
+    if (args_length/2 > server.hash_max_listpack_entries) {
+        hashTypeConvert(o, OBJ_ENCODING_HT);
+        return;
+    }
```

Comment:
some could be dups (which is also why we don't add the ones currently already in the dict).
but i suppose that the chances someone is adding many duplicate items are low, and the gain of not doing excessive work on a listpack is big.

for some reason i vaguely remember that we wanted to add such a thing in the past and dismissed that, but if it was important, i imagine we would have added a test that would fail.

the only thing missing IMHO is to add a similar change in SADD

Reply:
@madolson reminder to look into a similar improvement for sets

Reply:
Sorry for the delay. The reason I didn't add it here originally is that sets are sometimes encoded as intsets, and intset->hash is a one way conversion, so we would need to change the validation to also check if all of the values are integers if it's still a listpack. If we are okay with that I can add that too.

Reply:
i'm not sure i understand what that has to do with the topic of this PR.
i.e. what we do here is that on mass insertion, we check if the number of new items is over the packed limit, and then convert it to a hash at the beginning.
so you can do that comparison against either `set_max_intset_entries` or `set_max_listpack_entries` (depending on the current encoding).

we can maybe go a bit further and propagate a `future_count` hint from `saddCommand` down to `setTypeAddAux`, so that if an intset is converted to listpack, and we can predict that it's gonna be soon converted to dict, we convert it to dict directly.

Reply:
The gap I see is I was expecting this mostly for **new** objects, not mass insertions into existing objects. In that case, we don't know the existing encoding unless we go and check the args. I'll see if I can come up with some compromise here.

Reply:
your existing changes in hash are the same get triggered on existing keys as well. and as i noted in my initial comment, we don't count the existing fields since some of the new ones could be overwrites.
i don't see how is is logically different in sets (just a little bit more complicated to implement)

Reply:
Done.

Reply:
what you did is nice, but can't we also handle existing sets better?
i.e when user adds 500 members to an existing set that had 5?
either by passing a hint to setTypeAddAux? or by calling some method before the loop same as hsetCommand calls hashTypeTryConversion?